# Segmenting and Clustering Neighborhoods in Toronto

## Step 1: Scrape the following Wikipedia page and create a table

In [2]:
# import libraries
from bs4 import BeautifulSoup
import requests
import pandas as pd

In [15]:
# download url data from internet
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(url).text
Canada_data = BeautifulSoup(source, 'lxml')

In [16]:
# creat a new Dataframe
column_names = ['Postalcode','Borough','Neighborhood']
toronto = pd.DataFrame(columns = column_names)

In [60]:
# loop through to find postcode, borough, neighborhood 
content = Canada_data.find('div', class_='mw-parser-output')
table = content.table.tbody
postcode = 0
borough = 0
neighborhood = 0

for tr in table.find_all('tr'):
    i = 0
    for td in tr.find_all('td'):
        if i == 0:
            postcode = td.text
            i = i + 1
        elif i == 1:
            borough = td.text
            i = i + 1
        elif i == 2: 
            neighborhood = td.text.strip('\n').replace(']','')
    toronto = toronto.append({'Postalcode': postcode,'Borough': borough,'Neighborhood': neighborhood},ignore_index=True)

# clean dataframe 
toronto = toronto[toronto.Borough!='Not assigned']
toronto = toronto[toronto.Borough!= 0]
toronto.reset_index(drop = True, inplace = True)
i = 0
for i in range(0,toronto.shape[0]):
    if toronto.iloc[i][2] == 'Not assigned':
        toronto.iloc[i][2] = toronto.iloc[i][1]
        i = i+1
                                 
df = toronto.groupby(['Postalcode','Borough'])['Neighborhood'].apply(', '.join).reset_index()

def neighborhood_list(grouped):    
    return ', '.join(sorted(grouped['Neighborhood'].tolist()))
                    
grp = df.groupby(['Postalcode', 'Borough'])
df2 = grp.apply(neighborhood_list).reset_index(name='Neighborhood')
df2.rename(columns={'Postalcode':'Postal Code'}, inplace=True)
df2.head()

Postal Code      Borough                                       Neighborhood
0         M1B  Scarborough                     Rouge, Malvern, Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union, Highla...
2         M1E  Scarborough  Guildwood, Morningside, West Hill, Guildwood, ...
3         M1G  Scarborough                                     Woburn, Woburn
4         M1H  Scarborough                               Cedarbrae, Cedarbrae

Data cleaning

In [61]:
print(df2.shape)
df2.head(10)

(103, 3)


Postal Code      Borough                                       Neighborhood
0         M1B  Scarborough                     Rouge, Malvern, Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union, Highla...
2         M1E  Scarborough  Guildwood, Morningside, West Hill, Guildwood, ...
3         M1G  Scarborough                                     Woburn, Woburn
4         M1H  Scarborough                               Cedarbrae, Cedarbrae
5         M1J  Scarborough           Scarborough Village, Scarborough Village
6         M1K  Scarborough  East Birchmount Park, Ionview, Kennedy Park, E...
7         M1L  Scarborough  Clairlea, Golden Mile, Oakridge, Clairlea, Gol...
8         M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village Wes...
9         M1N  Scarborough  Birch Cliff, Cliffside West, Birch Cliff, Clif...

In [62]:
print('The DataFrame shape is', df2.shape)

The DataFrame shape is (103, 3)


The dataframe has 103 Postal codes but it has 212 rows, because each Postal code can present more than one neighborhood (210 in total). Therefore, the dataframe should be group by the Postal code, ending with a dataframe with 103 rows.

## Step 2:  Add the latitude and longitude coordinates to the dataframe

In [63]:
# For this task, we just use a prepared csv to retrieve the coordinates. Extract csv with Toronto geographical coordinates to dataframe.
toronto_geocsv = 'https://cocl.us/Geospatial_data'
!wget -q -O 'toronto_m.geospatial_data.csv' toronto_geocsv
geocsv_data = pd.read_csv(toronto_geocsv).set_index("Postal Code")
geocsv_data.head()

Latitude  Longitude
Postal Code                      
M1B          43.806686 -79.194353
M1C          43.784535 -79.160497
M1E          43.763573 -79.188711
M1G          43.770992 -79.216917
M1H          43.773136 -79.239476

In [64]:
# Combine the two tables

df = pd.merge(geocsv_data, df2, on='Postal Code')
df.head()

Postal Code   Latitude  Longitude      Borough  \
0         M1B  43.806686 -79.194353  Scarborough   
1         M1C  43.784535 -79.160497  Scarborough   
2         M1E  43.763573 -79.188711  Scarborough   
3         M1G  43.770992 -79.216917  Scarborough   
4         M1H  43.773136 -79.239476  Scarborough   

                                        Neighborhood  
0                     Rouge, Malvern, Rouge, Malvern  
1  Highland Creek, Rouge Hill, Port Union, Highla...  
2  Guildwood, Morningside, West Hill, Guildwood, ...  
3                                     Woburn, Woburn  
4                               Cedarbrae, Cedarbrae

In [65]:
df = df[['Postal Code', 'Borough', 'Neighborhood', 'Latitude', 'Longitude']]
df.head(11)

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   

                                         Neighborhood   Latitude  Longitude  
0                      Rouge, Malvern, Rouge, Malvern  43.806686 -79.194353  
1   Highland Creek, Rouge Hill, Port Union, Highla...  43.784535 -79.160497  
2   Guildwood, Morningside, West Hill, Guildwood, ...  43.763573 -79.188711  
3                                      Woburn, Woburn  43.770992 -79.216917  
4                                Cedarbrae, Cedarbrae  43.773136 -79.239476  
5            Scarborough Village, Scarborough Village  43.744734 -79.239476  
6   East Birchmount Park, Ionview, Kennedy Park, E...  43.727929 -79.262029  
7   Clairlea, Golden Mile, Oakridge, Clairlea, Gol...  43.711112 -79.284577  
8   Cliffcrest, Cliffside, Scarborough Village Wes...  43.716316 -79.239476  
9   Birch Cliff, Cliffside West, Birch Cliff, Clif...  43.692657 -79.264848  
10  Dorset Park, Scarborough Town Centre, Wexford ...  43.757410 -79.273304

In [69]:
df.to_csv('Toronto.TASK_II_df.csv',index=False)

In [72]:
df.shape

(103, 5)

## Step 3:  Exploring and clustering the neighborhoods in Toronto-part 3


In [73]:
#!conda install -c conda-forge geopy --yes        # if needed
#!conda install -c conda-forge folium=0.5.0 --yes # if needed

import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files


from geopy.geocoders import Nominatim 
GeoLocator = Nominatim(user_agent='My-IBMNotebook')# convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    folium-0.5.0               |             py_0          45 KB  conda-forge
    openssl-1.1.1d             |       h516909a_0         2.1 MB  conda-forge
    altair-4.0.1               |             py_0         575 KB  conda-forge
    ca-certificates-2019.11.28 |       hecc5488_0         145 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    certifi-2019.11.28         |   py36h9f0ad1d_1         149 KB  conda-forge
    branca-0.4.0               |             py_0          26 KB  conda-forge
    ------------------------------------------------------------
                       

In [74]:
toronto_task2_csv = "Toronto.TASK_II_df.csv"

In [75]:
toronto_neighborhoods = pd.read_csv(toronto_task2_csv)

In [77]:
toronto_neighborhoods.shape
toronto_neighborhoods.head()

Postal Code      Borough                                       Neighborhood  \
0         M1B  Scarborough                     Rouge, Malvern, Rouge, Malvern   
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union, Highla...   
2         M1E  Scarborough  Guildwood, Morningside, West Hill, Guildwood, ...   
3         M1G  Scarborough                                     Woburn, Woburn   
4         M1H  Scarborough                               Cedarbrae, Cedarbrae   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

Use geopy library to get the latitude and longitude values of Toronto Canada

In [78]:
address = 'Toronto, Ontario Canada'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto Canada are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto Canada are 43.653963, -79.387207.


/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


Create a map of Toronto with neighborhoods superimposed on top

In [80]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_neighborhoods['Latitude'], toronto_neighborhoods['Longitude'], toronto_neighborhoods['Borough'], toronto_neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=4,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#87cefa',
        fill_opacity=0.5,
        parse_html=False).add_to(map_toronto)
    
map_toronto

For this task, I will just focus the scope of analysis to neighborhoods in East,West and Central Toronto only.

In [81]:
toronto_data = toronto_neighborhoods[toronto_neighborhoods['Borough'].str.contains("Toronto")].reset_index(drop=True)
print(toronto_data.shape)
toronto_data.head()

(39, 5)


Postal Code          Borough  \
0         M4E     East Toronto   
1         M4K     East Toronto   
2         M4L     East Toronto   
3         M4M     East Toronto   
4         M4N  Central Toronto   

                                        Neighborhood   Latitude  Longitude  
0                           The Beaches, The Beaches  43.676357 -79.293031  
1  The Danforth West, Riverdale, The Danforth Wes...  43.679557 -79.352188  
2  The Beaches West, India Bazaar, The Beaches We...  43.668999 -79.315572  
3                   Studio District, Studio District  43.659526 -79.340923  
4                       Lawrence Park, Lawrence Park  43.728020 -79.388790

Re-create the map with new markers for Toronto Neighborhoods

In [82]:

# I will be using the same coordinates for the previous view
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Utilizing the Foursquare API to explore and segment neighborhoods

In [84]:
CLIENT_ID = 'JQPY5CEXAS5NJSNCGEO0E5BIT5Z23GYNW5SDWZPHVUCQFFBS' # your Foursquare ID
CLIENT_SECRET = 'UAAQGNUE02I23Y0XJJUT052TVDWHHBN10YQWAIGBCK41P0JX' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: JQPY5CEXAS5NJSNCGEO0E5BIT5Z23GYNW5SDWZPHVUCQFFBS
CLIENT_SECRET:UAAQGNUE02I23Y0XJJUT052TVDWHHBN10YQWAIGBCK41P0JX


In [86]:
# The following function retrieves the venues given the names and coordinates and stores it into dataframe.
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [89]:
toronto_venues = getNearbyVenues(names=toronto_neighborhoods['Neighborhood'],
                                   latitudes=toronto_neighborhoods['Latitude'],
                                   longitudes=toronto_neighborhoods['Longitude'])

Rouge, Malvern, Rouge, Malvern
Highland Creek, Rouge Hill, Port Union, Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill, Guildwood, Morningside, West Hill
Woburn, Woburn
Cedarbrae, Cedarbrae
Scarborough Village, Scarborough Village
East Birchmount Park, Ionview, Kennedy Park, East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge, Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West, Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West, Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights, Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford, Maryvale, Wexford
Agincourt, Agincourt
Clarks Corners, Sullivan, Tam O'Shanter, Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East, Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West, L'Amoreaux West
Upper Rouge, Upper Rouge
Hillcrest Villa

In [90]:
print(toronto_venues.shape)
toronto_venues.head()

(1332, 7)


Neighborhood  Neighborhood Latitude  \
0                     Rouge, Malvern, Rouge, Malvern              43.806686   
1                     Rouge, Malvern, Rouge, Malvern              43.806686   
2  Highland Creek, Rouge Hill, Port Union, Highla...              43.784535   
3  Guildwood, Morningside, West Hill, Guildwood, ...              43.763573   
4  Guildwood, Morningside, West Hill, Guildwood, ...              43.763573   

   Neighborhood Longitude                  Venue  Venue Latitude  \
0              -79.194353                Wendy's       43.807448   
1              -79.194353  Interprovincial Group       43.805630   
2              -79.160497  Royal Canadian Legion       43.782533   
3              -79.188711      G & G Electronics       43.765309   
4              -79.188711             Marina Spa       43.766000   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.200378            Print Shop  
2       -79.163085                   Bar  
3       -79.191537     Electronics Store  
4       -79.191000                   Spa